### PR TITLE
Adding SilentTeleport fix.

### PR DESCRIPTION
**Pull request by RogueDrifter**
Added SilentTeleport fixes for when someone tries to car-jack someone as his vehicle teleports.

I think I've done everything right and I tried to keep it as simple as possible, but, take a look and let me know, I was confused a bit about the loops so I used simple ones under OnPlayerEnterVehicle and SetVehiclePos hooked function, also tagged you about it on discord @Y-Less 

Original thread: http://forum.sa-mp.com/showthread.php?t=650422
Issue: #92 @NexiusTailer

### CODE REVIEW

File: fixes.inc [hunk 1]
```diff
@@ -5142,6 +5169,52 @@ static _FIXES_DetermineOS()
...
+native BAD_SetVehiclePos(vehicleid, Float:x, Float:y, Float:z) = SetVehiclePos; 
+
+/**
+ * <fixes>SilentTeleport</fixes>
+ */
+
+#if FIX_SilentTeleport
+     stock FIXES_SetVehiclePos(vehicleid, Float:x, Float:y, Float:z)
+     { 
+		 for(new i; i < MAX_PLAYERS; i++)
```
Comment (ziggi) [inline fixes.inc]: You need to use _FIXES_FOREACH loop instead

File: fixes.inc [hunk 2]
```diff
@@ -5142,6 +5169,52 @@ static _FIXES_DetermineOS()
...
+		 {
+			 if(!IsPlayerConnected(i)) continue;
+			 if(GetPlayerState(i) == PLAYER_STATE_DRIVER && GetPlayerVehicleID(i) == vehicleid)
+			 {
+				 if(FIXES_gsJackedId(i) != INVALID_PLAYER_ID)
```
Comment (ziggi) [inline fixes.inc]: This should be with square brackets

File: fixes.inc [hunk 3]
```diff
@@ -5142,6 +5169,52 @@ static _FIXES_DetermineOS()
...
+				 {
+					 ClearAnimations(FIXES_gsJackedId[i]);
+					 break;
+				 }
+			 }
+		 }
+		 SetVehiclePos(vehicleid, x, y, z);
+         return 0;
```
Comment (ziggi) [inline fixes.inc]: Need to return SetVehiclePos value.

File: fixes.inc [hunk 4]
```diff
@@ -5142,6 +5169,52 @@ static _FIXES_DetermineOS()
...
+					 break;
+				 }
+			 }
+		 }
+		 SetVehiclePos(vehicleid, x, y, z);
+         return 0; 
+     } 
+
+     #define _ALS_SetVehiclePos
+     #define SetVehiclePos( FIXES_SetVehiclePos(
```
Comment (ziggi) [inline fixes.inc]: Also what if the position of the vehicle will change on small value? I think what we need to add distance check.

**Review by ziggi — CHANGES_REQUESTED** [review]
I'm not sure, but maybe this fix can be merged with existing PassengerSeating fix, need to research.

Also don't forget about single space between operator and bracket ;)

File: fixes.inc (hunk 4 above)
Comment (RogueDrifter) [reply fixes.inc]: The player shouldn't teleport regardless of the distance of the vehicle teleporting as that teleport is 'silent' as in not recorded anywhere and can cause many problems related to anti-cheats and so on.

File: fixes.inc (hunk 3 above)
Comment (RogueDrifter) [reply fixes.inc]: Many other hooks were not using this method so i thought it wouldn't matter much, nevertheless changed it.

File: fixes.inc (hunk 2 above)
Comment (RogueDrifter) [reply fixes.inc]: You're right my bad, changed that.

File: fixes.inc (hunk 1 above)
Comment (RogueDrifter) [reply fixes.inc]: You should check discord i tagged u as well about this, regardless i've changed it as well.

**Comment by RogueDrifter** [comment]
And about passenger seating, i wouldn't know as i haven't applied that 'fix' and the file was confusing as it is, I've changed the requested changes.

File: fixes.inc [hunk 5]
```diff
@@ -5142,6 +5169,50 @@ static _FIXES_DetermineOS()
...
+native BAD_SetVehiclePos(vehicleid, Float:x, Float:y, Float:z) = SetVehiclePos; 
+
+/**
+ * <fixes>SilentTeleport</fixes>
+ */
+
+#if FIX_SilentTeleport
+	stock FIXES_SetVehiclePos(vehicleid, Float:x, Float:y, Float:z)
+    {	 
+		_FIXES_FOREACH(FIXES_gsJackedId, i)
```
Comment (ziggi) [inline fixes.inc]: It's wrong, `_FIXES_FOREACH` works differently, just use `_FIXES_FOREACH(FIXES_gsPlayersIterator, i)` for this.

File: fixes.inc [hunk 6]
```diff
@@ -6836,6 +6917,16 @@ native BAD_SetPlayerName(playerid, const name[]) = SetPlayerName;
 		//  END:   TogglePlayerControllable
 		// =================================
 
+	  	// =========================
+		// BEGIND: SilentTeleport
+		// =========================
+		#if FIX_SilentTeleport
```
Comment (ziggi) [inline fixes.inc]: You forgot to add check for definition of `FIX_SilentTeleport` before `OnPlayerSpawn`.

File: fixes.inc [hunk 7]
```diff
@@ -7707,6 +7812,27 @@ native BAD_TogglePlayerControllable(playerid, toggle) = TogglePlayerControllable
 		//  END: PassengerSeating
 		// =============================
 
+		// =============================
+		//  BEGIN: SilentTeleport
+		// =============================
+		#if FIX_SilentTeleport
```
Comment (ziggi) [inline fixes.inc:7825]: Same, you forgot to add check for definition of `FIX_SilentTeleport` before `OnPlayerEnterVehicle`.

**Review by ziggi — CHANGES_REQUESTED** [review]
@Y-Less can you check this too?

File: fixes.inc (hunk 6 above)
Comment (RogueDrifter) [reply fixes.inc]: Done.

File: fixes.inc (hunk 7 above)
Comment (RogueDrifter) [reply fixes.inc:7825]: Done.

File: fixes.inc (hunk 5 above)
Comment (RogueDrifter) [reply fixes.inc]: Actually if you don't mind explaining exactly how FIXES_FOREACH works that'd be great, i also think it should be explained in the style section of the documentation, many things are left out. Changed it.

**Comment by RogueDrifter** [comment]
One last comment as well, the fact that you told me to return SetVehiclePos instead when the style section of the documentation says to return 0; ??

Maybe the documentation should be more clear I assume? Anyhow I think I've got things right this time.

**Comment by Y-Less** [comment]
Returning the value from the native just ensures it is always correct, regardless of how correct the documentation may be, or how the native may change - it just saves effort and worry for no cost. In fact, many natives return 1 or 0 and can be used in place of IsPlayerConected (not that one though), it often makes sense to call the native first.

I was only going to comment on the missing FIX_ checks for callbacks, but you already got that, so great. Otherwise, it is looking very good - thank you very much for the contribution and putting up with what I know is (unfortunately necessarily) complex code.

Two questions:

1) How does this behave with filterscripts?
2) Would it be better to store vehicle drivers in an array (that's just an optional optimisation though).

**Comment by NexiusTailer** [comment]
I just wanted to recommend a different way to call the original natives from hook, but it seems Y_less is already described my thoughts.

Hovewer I would like to give an example of code that I personally find the most optimal solution:
```
stock FIXES_SetVehiclePos(vehicleid, Float:x, Float:y, Float:z)
{
	if(SetVehiclePos(vehicleid, x, y, z) == 0) return 0; //We don't need any other checks for validity of vehicleid or smth else as original native returns 0 when something wrong
	//Code
	return 1; //All is ok
}
```

**Comment by Y-Less** [comment]
That only works if the native doesn't always return 0, as I think this one might.

**Comment by NexiusTailer** [comment]
As I personally tested before, SetVehiclePos returns 1 if vehicle is valid and 0 if not. It doesn't check any other params except the first one, but I think that's enough. So there are also a few functions that always returns the same values like ChangeVehiclePaintjob or SetVehicleToRespawn, but the most of functions tends to normal logic 1/0.

**Comment by RogueDrifter** [comment]
I'm not entirely sure what's requested anymore but for those questions:
1- It doesn't matter if It's a Filterscript or a gamemode, it doesn't matter if its included in both or included just once.

2- It just adds more code, if you'll store the vehicle drivers in an array i assume a constant loop to always catch the new drivers of the vehicles and whether they're in them or not, it's up to you to apply that if you think it'll be an optimization but to me it doesn't make much of a difference.

And are you trying to say I should put up a check to see if(!IsValidVehicle(vehicleid)) return 0; about that other comment underneath SetVehiclePos function? (Or check if it returns 0 as a whole function i think both are the same since you're saying it returns 0 if the vehicle isn't valid).

**Review by Y-Less — APPROVED** [review]

**Comment by Y-Less** [comment]
I added the code we were talking about with using `SetVehiclePos` as the validity check, and approved the merge.  We can always optimise to remove the loop later.

**Pull request merged**